index/template/field_iterator.go

```diff
@@ -0,0 +1,60 @@
+package template
+
+import (
+	"github.com/xichen2020/eventdb/index"
+
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+type valueFieldIterator struct {
+	docIt index.DocIDSetIterator
+	valIt ForwardValueIterator
+
+	done      bool
+	err       error
+	currDocID int32
+	currValue GenericValue
+}
+
+func newValueFieldIterator(
+	docIt index.DocIDSetIterator,
+	valIt ForwardValueIterator,
+) *valueFieldIterator {
+	return &valueFieldIterator{
+		docIt: docIt,
+		valIt: valIt,
+	}
+}
+
+func (it *valueFieldIterator) Next() bool {
+	if it.done || it.err != nil {
+		return false
+	}
+	if !it.docIt.Next() || !it.valIt.Next() {
+		it.done = true
+		// TODO(xichen): Add doc it errors as well.
```

Comment:
this seems trivial, why punt?

Reply:
Yeah...except that the Err() method hasn’t been added to the DocIDSetIterator interface yet. I have a TODO for that so I’ll address this as well in the same PR.